### ISSUE TITLE
ignore instrumental versions of songs

### ISSUE THREAD

**Issue by skoef** [raveberry/raveberry#71]
First a little context: the way I use raveberry is that I have a simple daemon listening to a icecast stream with the sole purpose of detecting song changes. When a new song comes on, it requests this song on raveberry through the `/musiq/post_song/` API. This seems redundant, but it allows me and my colleagues to create our own playlist and if nobody feels like queueing music, that particular stream makes sure the playlist is always filled. I'm happy to share the source code for this monstrosity if anyone is interested.

Anyway, everytime this daemon requests `busta rhymes ft mariah carey - i know what you want` (please don't judge) it will queue the same *instrumental* version of the song (`I Know What You Want (In the Style of Busta Rhymes ft. Mariah Carey) [Instrumental Backing Track] (spotify:track:06FyURzp4xS1AQpjNdp3Gt)` in this case).

I'm not sure which code-path is responsible for picking the song out of the results returned by one or more providers, but could we ignore instrumental (and karaoke) versions of songs? Or perhaps make it a setting? If @raveberry does a suggestion where this filtering could be applied, I'm happy to implement it myself!

**Comment by raveberry** [comment]
I like your solution to always keep the queue filled. There is the autoplay option, but it tends to hover around the same songs over and over again. So yes, I am interested in your code :)

The filtering sounds like a good idea. I could imagine a comma separated list of keywords that could be specified in the settings, and any song containing a word from this list is disregarded.

tl;dr: use the suggestion at the bottom.

## A short digression into the song request cycle of Raveberry

Implementing it will be a little bit more complicated though, I'm afraid. Let me explain:

The request starts in `musiq.py:do_request_music`. Each active provider is created and stored in a list (spotify, soundcloud and youtube each have their own provider). If the user has a platform preference, that will be the first one. Now the providers receive the query one after another. The first provider to successfully find a song will break the loop and the request is finished.

The process of providing a song can be seen in `music_provider.py:request`. If the song is not cached (file present on disk), it is checked for availability (are there any search results?). If the provider can not find any song, a ProviderError is raised, causing the next provider to try and find it. Is a song available, it needs to be made available and enqueued. Since making the song available can take some time (especially in the case of youtube, where it needs to be downloaded first), it is done in the background. During the synchronous request handling only a placeholder is enqueued that does not contain any actual information. If the background thread is finished, this placeholder is filled with the required info (url, title, artist etc.).

This process could probably be simplified for spotify, since checking availability and making a song available is the same thing: performing an api call. However, I wanted to have an interface that works for all providers, including the slower Youtube one. That is the reason why every provider uses placeholders, even if they aren't even visible on the web page.

Now (finally) to the problem with the filtering mechanism: The metadata is only known after the background thread finishes. By that time, the original request looping through all providers already returned. This means that we can not simply check the title of the song and raise a ProviderError if it matches the filterlist.

Solving this from the top level only has these two options, I think:

* Filter the song after the metadata is available. If the song is filtered, initiate a new request that carries the additional information that this specific provider should not be used. This would solve the issue without changing every specific provider, but will probably throw away a lot of otherwise good results that just were not considered.

* Rewrite the whole request cycle to allow for synchronous filtering. A lot of work I would definitely not recommend.

## Recommended solution

While writing the text above, it came to my mind that filtering is best implemented in the providers themselves. But now I wrote all of the text and don't want to delete it anymore, so it's going to stay.

Approach it this way: filter the song in every provider separately. Go into `spotify.py:gather_metadata` and change the `limit` key to a higher value to receive more search results. Then go through the results and skip the ones you don't like, preferably with a customizable list of keywords from the settings.
In the other providers the search results need to be handled similarly, every time in the `check_available` section.

Unfortunately, I have some real life stuff going on currently, so I'm not sure when I will be able to implement this for every provider the way I imagine it.
I would recommend you to simply change the spotify provider to filter out instrumental songs and you should be good to go for now.

**Comment by raveberry** [comment]
The filter function is now part of Raveberry with 0.7.3.
It does not work for Youtube though, as youtube-dl does not provide this functionality and I try to keep interaction with with Youtube directly to a minimum.
